File: 백준/Gold/13424. 비밀 모임/비밀 모임.py
def solve():
    n, m = map(int, input().split())
    dist = [[987654321 for _ in range(n+1)] for _ in range(n+1)]    
    for i in range(1, n+1):
        dist[i][i] = 0
    
    for _ in range(m):
        a,b,c = map(int, input().split())
        dist[a][b] = min(dist[a][b], c)
        dist[b][a] = min(dist[b][a], c)
    for k in range(1, n+1):
        for i in range(1, n+1):
            for j in range(1, n+1):
                dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
    k = int(input())
    friend = [*map(int, input().split())]                
    sum = 987654321
    ret = -1
    for i in range(1, n+1):
        tmp = 0
        for f in friend:            
            tmp = tmp + dist[f][i]
        if tmp < sum:
            sum = tmp
            ret = i
    print(ret)

        
    
    
t = int(input())
for _ in range(t):
    solve()

File: 백준/Gold/13424. 비밀 모임/비밀 모임.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static int dist[][];
    static boolean checked[];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            solve(br);
        }
    }

    public static void solve(BufferedReader br) throws IOException {
        String str = br.readLine();
        StringTokenizer s = new StringTokenizer(str);
        int n = Integer.parseInt(s.nextToken());
        int m = Integer.parseInt(s.nextToken());
        dist = new int[n + 1][n + 1];
        checked = new boolean[n + 1];
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                dist[i][j] = 987654321;
            }
            dist[i][i] = 0;
            checked[i] = false;
        }
        for (int i = 0; i < m; i++) {
            str = br.readLine();
            s = new StringTokenizer(str);
            int a = Integer.parseInt(s.nextToken());
            int b = Integer.parseInt(s.nextToken());
            int c = Integer.parseInt(s.nextToken());
            dist[a][b] = Math.min(dist[a][b], c);
            dist[b][a] = Math.min(dist[b][a], c);
        }
        for (int k = 1; k <= n; k++) {
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
        str = br.readLine();
        int k = Integer.parseInt(str);
        str = br.readLine();
        s = new StringTokenizer(str);
        for (int i = 0; i < k; i++) {
            int num = Integer.parseInt(s.nextToken());
            checked[num] = true;
        }
        int sum = 987654321;
        int idx = -1;
        for (int i = 1; i <= n; i++) {
            int tmp = 0;
            for (int j = 1; j <= n; j++) {
                if (checked[j]) {
                    tmp += dist[j][i];
                }

            }
            if (tmp < sum) {
                sum = tmp;
                idx = i;
            }
        }
        System.out.println(idx);
    }
}